ssh/tailssh/incubator.go

```diff
@@ -670,14 +670,31 @@ func (ss *sshSession) launchProcess() error {
 
 	cmd := ss.cmd
 	homeDir := ss.conn.localUser.HomeDir
-	if _, err := os.Stat(homeDir); err == nil {
-		cmd.Dir = homeDir
-	} else if os.IsNotExist(err) {
-		// If the home directory doesn't exist, we can't chdir to it.
-		// Instead, we'll chdir to the root directory.
+	// We prefer to execute the child process in the connecting user's home
+	// directory. We need to be able to chdir to that directory in order to do
+	// so. How to check this? We don't want to chdir the parent tailscaled
+	// process, because that could have unexpected side-effects depending on
+	// what assumptions tailscaled makes about its working directory.
+	//
+	// Instead, we assume that if we're able to read the directory, we can
+	// chdir there, and conversely if we can't read the directory, we can't
+	// chdir there.
+	//
+	// To test this, we attempt to read some of the contents of the directory.
+	// If this succeeds, we assume that we'll be able to chdir to the directory,
+	// otherwise we assume that we can't.
+	f, err := os.Open(homeDir)
+	if err != nil {
 		cmd.Dir = "/"
 	} else {
-		return err
+		if _, err := f.ReadDir(1); err == nil || err == io.EOF {
+			cmd.Dir = homeDir
+		} else {
+			cmd.Dir = "/"
+		}
+		if err := f.Close(); err != nil {
+			ss.logf("failed to close user homedir: %s", err)
+		}
```

Comment:
Can you pull this all out into a function so launchProcess isn't interrupted with this big detour in the middle of it?

Like:

```go
      if ss.dirIsReadable(homeDir) {
             cmd.Dir = homeDir
      } else {
             cmd.Dir = "/"
      }
```

Then you can `defer f.Close()` in the helper func (more idiomatic). And BTW, no need to check Close errors for files opened read-only. Only writable file handles (as they might not have flushed until Close).

Then you also move most of the new docs up to the dirIsReadable func. Or name it `useDirAsHomeDir` if you prefer.

Reply:
Done (named it `canChdirTo` though).

Reply:
Not a huge fan of that name, but at least it's unexported. I think `canChdirTo` is misleading because that's not what it it reports. It's reporting a guess as to whether we can chdir to it later. I'd be worried a future change refactors "can chdir to" as doing a chdir back and forth. Whereas `useDirAsHomeDir` says the caller's intent, and `dirIsReadable` says exactly what it reports.